internal/k8s/statusupdater.go

```diff
@@ -0,0 +1,186 @@
+// Copyright © 2020 VMware
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8s
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/projectcontour/contour/internal/assert"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// StatusUpdate contains an all the information needed to change an object's status to perform a specific update.
+// Send down a channel to the goroutine that actually writes the changes back.
+type StatusUpdate struct {
+	FullName FullName
+	GVR      schema.GroupVersionResource
+	Mutator  StatusUpdateMutator
+}
+
+// StatusUpdateMutator is an interface to hold mutator functions for status updates.
+type StatusUpdateMutator interface {
+	Mutate(obj interface{}) interface{}
+}
+
+// StatusUpdateHandler holds the details required to actually write an Update back to the referenced object.
+type StatusUpdateHandler struct {
+	Log           logrus.FieldLogger
+	Clients       *Clients
+	UpdateChannel chan StatusUpdate
+	LeaderElected chan struct{}
+	IsLeader      bool
+	Converter     *UnstructuredConverter
+}
+
+// Start runs the goroutine to perform status writes.
+// Until the Contour is elected leader, will drop updates on the floor.
+func (sw *StatusUpdateHandler) Start(stop <-chan struct{}) error {
+
+	for {
+		select {
+		case <-stop:
+			return nil
+		case <-sw.LeaderElected:
+			sw.Log.Info("elected leader")
+			sw.IsLeader = true
+			// disable this case
+			sw.LeaderElected = nil
+		case upd := <-sw.UpdateChannel:
+			if !sw.IsLeader {
+				sw.Log.WithField("fullname", upd.FullName).Debug("not leader, not applying update")
```

Comment:
In other places we use separate `name` and `namespace` fields. Not sure what is more useful, but we should be consistent here and below.

Reply:
Moved everything to `name` and `namespace` fields.